### ISSUE TITLE
External CA: kubeconfig file check still expects ca.key

### ISSUE THREAD

**Issue by mitchellmaler** [kubernetes/kubeadm#1391]
## Is this a BUG REPORT or FEATURE REQUEST? 
Bug Report

## Versions

**kubeadm version** (use `kubeadm version`):
kubeadm version: &version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-01T20:05:53Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-01T20:08:12Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
- **OS** (e.g. from /etc/os-release): Centos 7
- **Kernel** (e.g. `uname -a`): Linux master0 3.10.0-957.1.3.el7.x86_64 #1 SMP Thu Nov 29 14:49:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

## What happened?

I have manually placed all the pki certs, sa cert, and ca's without the ca keys. Then I placed the conf files in the /etc/kubernetes directory. When running `kubeadm init` it instantly returns an error stating the ca.key doesn't exist when processing the kubeconfig files.
```
$ sudo kubeadm init --config /vagrant/config/clusterConfiguration.yaml
couldn't create a kubeconfig; the CA files couldn't be loaded: failed to load key: couldn't load the private key file /etc/kubernetes/pki/ca.key: open /etc/kubernetes/pki/ca.key: no such file or directory
``` 

## What you expected to happen?
I would expect it to verify the ca cert in the config files but without the ca.key when using an external ca. It shouldn't try loading the key since it doesn't exist.

## How to reproduce it (as minimally and precisely as possible)?
Generate the ca's, certs, and place the kubeconfig files and place them in the correct directories without the ca keys. Run kubeadm init and it tries to process the kubeconfig files with the key but fails cause it doesn't exist. 

## Anything else we need to know?
Looks like it has to do with https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go#L133 which is called from https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go#L360 the ValidateKubeconfigsForExternalCA method.

**Comment by RA489** [comment]
I can look into it.

**Comment by yagonobre** [comment]
Thanks @RA489 
/lifecycle active

**Comment by linxcat** [comment]
This issue still needs a look. The pull request is currently insufficient.

**Comment by deblasis** [comment]
Hi @linxcat, since I am affected by the issue and I was monitoring for a
fix to be merged, I am happy to have a look into it instead. Please let me
know if I can

**Comment by linxcat** [comment]
@deblasis I'm not a maintainer. I'm in the same boat as you, I design architecture rather than write code. I'm hoping for this to be fixed so I'm not blocked behind a need to place the private key for a Certificate Authority on the root volume.

**Comment by linxcat** [comment]
Found [this](https://github.com/kubernetes/kubernetes/pull/75431) Which seems to indicate this is a very recent and live problem. I've referenced the author in the PR, but it looks like this could use some effort on the part of release managers in terms of backports etc. The PR I linked is tied to Milestone 1.14, but it really needs to be implemented in at least 1.13

**Comment by neolit123** [comment]
@linxcat we didn't consider https://github.com/kubernetes/kubernetes/pull/75431 for backport to 1.13, because @fabriziopandini recently found the problem in the code. but it makes sense to me to backport.

/assign @fabriziopandini @timothysc 
WDYT?

**Comment by fabriziopandini** [comment]
@neolit123 yes. I think that https://github.com/kubernetes/kubernetes/pull/75431 fixes the issue
Let's sync in slack about how to proceed/how to test the backport before submission

**Comment by linxcat** [comment]
Is there any update on the release status of this fix in 1.14 and the backport status in 1.13?

**Comment by neolit123** [comment]
the fix is in `v1.15.0-alpha.1  v1.14.2-beta.0 v1.14.1 v1.14.1-beta.0 v1.14.0 v1.14.0-rc.1`
but we haven't done the 1.13 backport yet.

**Comment by neolit123** [comment]
given the size of the change, we are probably going to hold on the 1.13 backport.
please upgrade to 1.14+.

**Comment by Venkatvenstvm** [comment]
Hi All -  Is this issue actually fixed i am using 1.18.6 version, with external CA, still it required CA.key when i am joining 2 nd controller node through load balancer with 1st controller.

**Comment by Venkatvenstvm** [comment]
> given the size of the change, we are probably going to hold on the 1.13 backport.
> please upgrade to 1.14+.

Hi All -  Is this issue actually fixed i am using 1.18.6 version, with external CA, still it required CA.key when i am joining 2 nd controller node through load balancer with 1st controller.

**Comment by neolit123** [comment]
@Venkatvenstvm hi, are you getting the same error message as above?

one problem about the missing ca.key is that even if you get kubeadm to start without it, it is still required by the kube-controller-manager to e.g. sign the CSRs for new joining nodes (kubelet client certs).

https://github.com/kubernetes/kubernetes/blob/af29f81af6f5afc0ad48cd2d3d5472a9ca4d4cc8/cmd/kubeadm/app/phases/controlplane/manifests.go#L346

**Comment by Venkatvenstvm** [comment]
> @Venkatvenstvm hi, are you getting the same error message as above?
> 
> one problem about the missing ca.key is that even if you get kubeadm to start without it, it is still required by the kube-controller-manager to e.g. sign the CSRs for new joining nodes (kubelet client certs).
> 
> https://github.com/kubernetes/kubernetes/blob/af29f81af6f5afc0ad48cd2d3d5472a9ca4d4cc8/cmd/kubeadm/app/phases/controlplane/manifests.go#L346

Hi Neolit123 - Thanks for your reply.. yes still getting the same error, i understood the functionality of controller manager that required key file to generate the new certificate, however is there any way to pass "cluster-signing-key-file" as empty string as mentioned in code (line-351) in run time without affecting current  master node..

**Comment by neolit123** [comment]
> however is there any way to pass "cluster-signing-key-file" as empty string as mentioned in code (line-351) in run time without affecting current master node..

you can try editing the manifest file to edit out the `cluster-signing-key-file` flag, but not sure how the component will behave with that. obviously it will not be able to sign CSRs any more.

in 1.18 and 1.19 has the --experimental-kustomize feature that can be used to customize the static pods with patches.
but we are replacing it in 1.20 with a slightly different method --experimental-patches.

> yes still getting the same error

the original error in this thread is about "kubeadm init" and i tested something related the other day that "ca.key" is not needed if you pass all kubeconfig/certs.

you've mentioned above:

> Hi All - Is this issue actually fixed i am using 1.18.6 version, with external CA, still it required CA.key when i am joining 2 nd controller node through load balancer with 1st controller.

please, try passing --v=5 to the "kubeadm join" command and show the output here.

**Comment by Venkatvenstvm** [comment]
> cluster-signing-key-file



> > however is there any way to pass "cluster-signing-key-file" as empty string as mentioned in code (line-351) in run time without affecting current master node..
> 
> you can try editing the manifest file to edit out the `cluster-signing-key-file` flag, but not sure how the component will behave with that. obviously it will not be able to sign CSRs any more.
> 
> in 1.18 and 1.19 has the --experimental-kustomize feature that can be used to customize the static pods with patches.
> but we are replacing it in 1.20 with a slightly different method --experimental-patches.
> 
> > yes still getting the same error
> 
> the original error in this thread is about "kubeadm init" and i tested something related the other day that "ca.key" is not needed if you pass all kubeconfig/certs.
> 
> you've mentioned above:
> 
> > Hi All - Is this issue actually fixed i am using 1.18.6 version, with external CA, still it required CA.key when i am joining 2 nd controller node through load balancer with 1st controller.
> 
> please, try pssing --v=5 to the "kubeadm join" command and show the output here.

Hi -  Sharing the output of kube join command with --v=5, i mean time trying other options mentioned by you..  

**Apologies** for confusing you with my answer that i am facing issue this only in join the another controller to the existing  not in init.  


I0922 06:05:02.675785   20058 join.go:388] [preflight] found /etc/kubernetes/admin.conf. Use it for skipping discovery
I0922 06:05:02.676589   20058 join.go:398] [preflight] found NodeName empty; using OS hostname as NodeName
I0922 06:05:02.676604   20058 join.go:402] [preflight] found advertiseAddress empty; using default interface's IP address as advertiseAddress
I0922 06:05:02.676638   20058 initconfiguration.go:103] detected and using CRI socket: /var/run/dockershim.sock
I0922 06:05:02.676741   20058 interface.go:400] Looking for default routes with IPv4 addresses
I0922 06:05:02.676748   20058 interface.go:405] Default route transits interface "ens5"
I0922 06:05:02.676822   20058 interface.go:208] Interface ens5 is up
I0922 06:05:02.676863   20058 interface.go:256] Interface "ens5" has 2 addresses :[172.31.38.173/20 fe80::8c5:acff:fe43:ed16/64].
I0922 06:05:02.676882   20058 interface.go:223] Checking addr  172.31.38.173/20.
I0922 06:05:02.676892   20058 interface.go:230] IP found 172.31.38.173
I0922 06:05:02.676900   20058 interface.go:262] Found valid IPv4 address 172.31.38.173 for interface "ens5".
I0922 06:05:02.676907   20058 interface.go:411] Found active IP 172.31.38.173 
[preflight] Running pre-flight checks
I0922 06:05:02.676962   20058 preflight.go:90] [preflight] Running general checks
I0922 06:05:02.676995   20058 checks.go:249] validating the existence and emptiness of directory /etc/kubernetes/manifests
I0922 06:05:02.677020   20058 checks.go:286] validating the existence of file /etc/kubernetes/kubelet.conf
	[WARNING FileAvailable--etc-kubernetes-kubelet.conf]: /etc/kubernetes/kubelet.conf already exists
I0922 06:05:02.677050   20058 checks.go:286] validating the existence of file /etc/kubernetes/bootstrap-kubelet.conf
I0922 06:05:02.677059   20058 checks.go:102] validating the container runtime
I0922 06:05:02.767314   20058 checks.go:128] validating if the "docker" service is enabled and active
	[WARNING IsDockerSystemdCheck]: detected "cgroupfs" as the Docker cgroup driver. The recommended driver is "systemd". Please follow the guide at https://kubernetes.io/docs/setup/cri/
I0922 06:05:02.854410   20058 checks.go:335] validating the contents of file /proc/sys/net/bridge/bridge-nf-call-iptables
I0922 06:05:02.854464   20058 checks.go:335] validating the contents of file /proc/sys/net/ipv4/ip_forward
I0922 06:05:02.854492   20058 checks.go:649] validating whether swap is enabled or not
I0922 06:05:02.854526   20058 checks.go:376] validating the presence of executable conntrack
I0922 06:05:02.854547   20058 checks.go:376] validating the presence of executable ip
I0922 06:05:02.854567   20058 checks.go:376] validating the presence of executable iptables
I0922 06:05:02.854586   20058 checks.go:376] validating the presence of executable mount
I0922 06:05:02.854604   20058 checks.go:376] validating the presence of executable nsenter
I0922 06:05:02.854620   20058 checks.go:376] validating the presence of executable ebtables
I0922 06:05:02.854636   20058 checks.go:376] validating the presence of executable ethtool
I0922 06:05:02.854659   20058 checks.go:376] validating the presence of executable socat
I0922 06:05:02.854675   20058 checks.go:376] validating the presence of executable tc
I0922 06:05:02.854691   20058 checks.go:376] validating the presence of executable touch
I0922 06:05:02.854712   20058 checks.go:520] running all checks
I0922 06:05:02.932009   20058 checks.go:406] checking whether the given node name is reachable using net.LookupHost
I0922 06:05:02.932153   20058 checks.go:618] validating kubelet version
I0922 06:05:02.983585   20058 checks.go:128] validating if the "kubelet" service is enabled and active
I0922 06:05:02.992144   20058 checks.go:201] validating availability of port 10250
I0922 06:05:02.992322   20058 checks.go:432] validating if the connectivity type is via proxy or direct
I0922 06:05:02.992372   20058 join.go:483] [preflight] Fetching init configuration
I0922 06:05:02.992380   20058 join.go:526] [preflight] Retrieving KubeConfig objects
[preflight] Reading configuration from the cluster...
[preflight] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
I0922 06:05:03.003394   20058 request.go:933] Got a Retry-After 1s response for attempt 1 to https://172.31.20.115:6443/api/v1/namespaces/kube-system/configmaps/kubeadm-config?timeout=10s
I0922 06:05:04.027026   20058 interface.go:400] Looking for default routes with IPv4 addresses
I0922 06:05:04.027041   20058 interface.go:405] Default route transits interface "ens5"
I0922 06:05:04.027141   20058 interface.go:208] Interface ens5 is up
I0922 06:05:04.027190   20058 interface.go:256] Interface "ens5" has 2 addresses :[172.31.38.173/20 fe80::8c5:acff:fe43:ed16/64].
I0922 06:05:04.027206   20058 interface.go:223] Checking addr  172.31.38.173/20.
I0922 06:05:04.027212   20058 interface.go:230] IP found 172.31.38.173
I0922 06:05:04.027226   20058 interface.go:262] Found valid IPv4 address 172.31.38.173 for interface "ens5".
I0922 06:05:04.027236   20058 interface.go:411] Found active IP 172.31.38.173 
I0922 06:05:04.120689   20058 preflight.go:101] [preflight] Running configuration dependant checks

One or more conditions for hosting a new control plane instance is not satisfied.

failure loading key for CA: couldn't load the private key file /etc/kubernetes/pki/ca.key: open /etc/kubernetes/pki/ca.key: no such file or directory

Please ensure that:
* The cluster has a stable controlPlaneEndpoint address.
* The certificates that must be shared among control plane instances are provided.


k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/join.runPreflight
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/join/preflight.go:114
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:234
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:422
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:207
k8s.io/kubernetes/cmd/kubeadm/app/cmd.NewCmdJoin.func1
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/join.go:174
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:842
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:950
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:887
k8s.io/kubernetes/cmd/kubeadm/app.Run
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
main.main
	_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
runtime.main
	/usr/local/go/src/runtime/proc.go:204
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1374
error execution phase preflight
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:235
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:422
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:207
k8s.io/kubernetes/cmd/kubeadm/app/cmd.NewCmdJoin.func1
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/join.go:174
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:842
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:950
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:887
k8s.io/kubernetes/cmd/kubeadm/app.Run
	/workspace/anago-v1.19.2-rc.0.12+19706d90d87784/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/app/kubeadm.go:50
main.main
	_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubeadm/kubeadm.go:25
runtime.main
	/usr/local/go/src/runtime/proc.go:204
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1374

**Comment by neolit123** [comment]
ok, i might need some more information to debug this please open a new issue and fill the issue template in *full detail*
thanks.
